### PR TITLE
use brotli to compress self profile response

### PR DESCRIPTION
**Pull request by s7tya**
alternative to #1966
This PR is using brotli to compress and consider Accept-Encoding header according to previous PR's discussion.

### CODE REVIEW

**Comment by s7tya** [comment]
It will take about 22s to load both directory giving gzip into the perfetto, and using brotli & content-encoding: brotli by my eyes & stopwatch.
However, it would be nice to use brotli than gzip if it's available, so I propose this than the previous one.

**Comment by s7tya** [comment]
(Current production RLO takes about 200s to load. so this makes x10 faster. )

**Comment by Kobzol** [comment]
Because we don't have https://github.com/rust-lang/rustc-perf/issues/1896 yet, it is not easy to fully test (and benchmark) this locally. So I guess we can try it on the live server and see how fast it is in practice.

File: site/src/server.rs [hunk 1]
```diff
@@ -440,8 +440,18 @@ async fn serve_req(server: Server, req: Request) -> Result<Response, ServerError
         }
         "/perf/processed-self-profile" => {
             let ctxt: Arc<SiteCtxt> = server.ctxt.read().as_ref().unwrap().clone();
+            let support_brotli = req
```
Comment (Kobzol) [inline site/src/server.rs:443]: There is already `allow_compression` on line 342, let's reuse it. This one has a bug; it should check `ACCEPT_ENCODING`, not `CONTENT_ENCODING` :) Splitting it by commas is nicer than just the `contains` check in `allow_compression`, but I think that in that case we should check if the trimmed string starts with `br`, because there can be some [compression options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding#syntax).

File: site/src/self_profile.rs [hunk 2]
```diff
@@ -37,9 +37,11 @@ pub fn generate(
         ProcessorType::Crox => {
             let opt = serde_json::from_str(&serde_json::to_string(&params).unwrap())
                 .context("crox opts")?;
+            let data = crox::generate(self_profile_data, opt).context("crox")?;
```
Comment (Kobzol) [inline site/src/self_profile.rs:40]: Let's revert this change to avoid an unnecessary git diff.

File: site/src/request_handlers/self_profile.rs [hunk 3]
```diff
@@ -132,6 +136,19 @@ pub async fn handle_self_profile_processed_download(
         })
         .status(StatusCode::OK);
 
+    if output.filename.ends_with("json") && support_brotli {
```
Comment (Kobzol) [inline site/src/request_handlers/self_profile.rs:139]: You could reuse `maybe_compressed_response` from the `server.rs` file. We just to check if the compression of the profiles is still good enough (i.e. that the resulting file is small enough) with the parameters configured in `server.rs`.

**Comment by Kobzol** [comment]
Looks good! :) Was `5` the best trade-off for the compression options that you found? :)

**Comment by s7tya** [comment]
I just took benchmark and changed to 4.

```
207.22 MB → 19.55 MB (size: 9.43%, quality: 2, time: 713.505083ms)
207.22 MB → 13.85 MB (size: 6.69%, quality: 3, time: 744.224166ms)
207.22 MB → 11.42 MB (size: 5.51%, quality: 4, time: 738.653667ms)
207.22 MB → 12.88 MB (size: 6.22%, quality: 5, time: 1.467692333s)
```

**Review by Kobzol — APPROVED** [review]
Thanks for doing the benchmark! The comment in `server.rs` said that `4` might be too slow, maybe on the production server it will be a bit worse, but I'm fine with starting with `4`, it looks like it has a good trade-off locally! It is surprising that `5` produces a larger output than `4` though.

Thanks for working on this :heart:
